unturned/OpenMod.Unturned/Users/UnturnedUserProvider.cs

```diff
@@ -63,8 +63,46 @@ public class UnturnedUserProvider : IUserProvider, IDisposable
             Provider.onEnemyConnected += OnPlayerConnected;
             Provider.onEnemyDisconnected += OnPlayerDisconnected;
             Provider.onRejectingPlayer += OnRejectingPlayer;
-        }
-
+            Provider.onBanPlayerRequested += OnBanRequested;
+        }
+
+        private void OnBanRequested(CSteamID instigator, CSteamID playerToBan, uint ipToBan, ref string reason, ref uint duration, ref bool shouldVanillaBan)
+        {
+            var user = GetUser(playerToBan)!;
+
+            string punisherId;
+            if (instigator == null)
+            {
+                punisherId = "0";
+            }
+            else
+            {
+                punisherId = instigator.ToString();
+            }
+
+            var banReason = reason;
+            var banTime = duration;
+            bool ban = shouldVanillaBan;
+
+            AsyncHelper.RunSync(async () =>
+            {
+                var @event = new UnturnedUserBanningEvent(user, punisherId, banReason, DateTime.Now.AddSeconds(banTime));
+
+                @event.IsCancelled = !ban;
+
+                await m_EventBus.EmitAsync(m_Runtime, this, @event);
+
+                if (@event.IsCancelled)
+                {
+                    ban = false;
+                }
+            });
+
+            reason = banReason;
+            duration = banTime;
+            shouldVanillaBan = ban;
+        }
+

```

Comment:
Should be removed. Use instead [BansEventsListener](https://github.com/openmod/openmod/blob/main/unturned/OpenMod.Unturned/Players/Bans/Events/BansEventsListener.cs#L37)